src/main/java/build/buf/protovalidate/CustomOverload.java

```diff
@@ -611,40 +656,583 @@ private static boolean validateURI(String val, boolean checkAbsolute) {
+    if (version == 6L) {
+      Ipv6 ip = new Ipv6(str);
+      return ip.addressPrefix() && (!strict || ip.isPrefixOnly());
+    } else if (version == 4L) {
+      Ipv4 ip = new Ipv4(str);
+      return ip.addressPrefix() && (!strict || ip.isPrefixOnly());
+    } else if (version == 0L) {
+      return isIPPrefix(str, 6, strict) || isIPPrefix(str, 4, strict);
+    }
+    return false;
+  }
+}
+
+final class Ipv4 {
+  private String str;
+  private int index;
+  private List<Short> octets;
+  private long prefixLen;
+
+  Ipv4(String str) {
+    this.str = str;
+    this.octets = new ArrayList<Short>();
+  }
+
+  /**
+   * Returns the 32-bit value of an address parsed through address() or addressPrefix().
+   *
+   * <p>Note Java does not support unsigned numeric types, so to handle unsigned 32-bit values, we
+   * need to use a 64-bit long type instead of the 32-bit (signed) Integer type.
+   *
+   * <p>Returns 0 if no address was parsed successfully.
+   */
+  int getBits() {
+    if (this.octets.size() != 4) {
+      return -1;
+    }
+    return (this.octets.get(0) << 24)
+        | (this.octets.get(1) << 16)
+        | (this.octets.get(2) << 8)
+        | this.octets.get(3);
+  }
+
+  /**
+   * Returns true if all bits to the right of the prefix-length are all zeros.
+   *
+   * <p>Behavior is undefined if addressPrefix() has not been called before, or has returned false.
    */
-  private static boolean validateIPPrefix(String prefix, long ver, boolean strict) {
-    IPAddressString str;
-    IPAddress addr;
+  boolean isPrefixOnly() {
+    int bits = this.getBits();
+
+    int mask = 0;
+    if (this.prefixLen == 32) {
+      mask = 0xffffffff;
+    } else {
+      mask = ~(0xffffffff >>> this.prefixLen) >>> 0;
+    }
+
+    int masked = (bits & mask) >>> 0;
+
+    return bits == masked;
+  }
+
+  // Parses an IPv4 Address in dotted decimal notation.
+  boolean address() {
+    return this.addressPart() && this.index == this.str.length();
+  }
+
+  // Parses an IPv4 Address prefix.
+  boolean addressPrefix() {
+    return this.addressPart()
+        && this.take('/')
+        && this.prefixLength()
+        && this.index == this.str.length();
+  }
+
+  // Stores value in `prefixLen`
+  private boolean prefixLength() {
+    int start = this.index;
+
+    while (true) {
+      if (this.index >= this.str.length() || !this.digit()) {
+        break;
+      }
+
+      if (this.index - start > 2) {
+        // max prefix-length is 32 bits, so anything more than 2 digits is invalid
+        return false;
+      }
+    }
+
+    String str = this.str.substring(start, this.index);
+    if (str.length() == 0) {
+      // too short
+      return false;
+    }
+
+    if (str.length() > 1 && str.charAt(0) == '0') {
+      // bad leading 0
+      return false;
+    }
+
     try {
-      str = new IPAddressString(prefix);
-      addr = str.toAddress();
-    } catch (Exception e) {
+      int val = Integer.parseInt(str);
+
```

Comment:
Lots of extra whitespace in here that isn't improving readability - I'd remove the majority of this.